lib/install/management/appliance.go

```diff
@@ -660,6 +691,96 @@ func (d *Dispatcher) createAppliance(conf *config.VirtualContainerHostConfigSpec
 	return nil
 }
 
+// createVCHInventoryFolders creates the default structure of the inventory for the VCH and returns the object reference to the last folder created
+func (d *Dispatcher) createVCHInventoryFolders(spec *types.VirtualMachineConfigSpec) (*object.Folder, error) {
+	d.op.Info("Creating VCH inventory folders")
+	vchName := spec.Name
+
+	// assemble creation list
+	folders := []string{d.session.Cluster.Name()}
+	consolidated := strings.TrimLeft(d.vchPoolPath, d.session.ClusterPath)
+	folders = append(folders, strings.Split(consolidated, "/")...)
+	d.op.Debugf("Determined inventory targets for creation: %s", folders)
+
+	// Find all of the folders that we can before we start creating.
+	existingFolder, foldersToCreate, err := d.findInventoryCreationShortlist(folders)
+	if err != nil {
+		return nil, err
+	}
+
+	d.op.Debugf("Determined that these folders need to be created: %s", foldersToCreate)
+
+	// now that we have the list of needed constructions we need to assemble it.
+	vchFolder, err := d.createFolderChainForVCH(foldersToCreate, existingFolder, vchName)
+	if err != nil {
+		return nil, err
+	}
+
+	return vchFolder, nil
+}
+
+func (d *Dispatcher) findInventoryCreationShortlist(folders []string) (*object.Folder, []string, error) {
+	var targetPath string
+	creationFolders := make([]string, 0)
+
+	dcfolder, err := d.session.Datacenter.Folders(d.op)
+	if err != nil {
+		return nil, nil, err
+	}
+	folder := dcfolder.VmFolder
+
+	for index, folderName := range folders {
+		targetPath = fmt.Sprintf("%s/%s", folder.InventoryPath, folderName)
+
+		tempRef, _ := d.session.Finder.Folder(d.op, targetPath)
+		if tempRef == nil {
+			// collect the rest of the folders and return them
+			for index < len(folders) {
+				creationFolders = append(creationFolders, folders[index])
+				index++
+			}
+			break
+		}
+		folder = tempRef
+	}
+	return folder, creationFolders, nil
+}
+
+// createFolderChain will create the list of folders as a chain(each one the parent of the next) starting at the give "startingFolder"
+// WARNING: this will attempt a cleanup at the base of the chain if it fails. do not use if you suspect files exists past startingFolder and the first file creation element.
+func (d *Dispatcher) createFolderChainForVCH(folders []string, startingFolder *object.Folder, vchName string) (*object.Folder, error) {
+
+	createdFolder := startingFolder
+	for _, fname := range folders {
+		folderCreateTarget := fmt.Sprintf("%s/%s", createdFolder.InventoryPath, fname)
+		d.op.Debugf("Attempting to create folder : %s", folderCreateTarget)
+
+		// attempt to create the folder
+		f, err := createdFolder.CreateFolder(d.op, fname)
+		err = processInventoryCreationError(d.op, err, fname, vchName)
+
+		// creation failed, we must cleanup and bail
+		if err != nil {
```

Comment:
I want to note for discussion sake, we do not check the error for a logical action here. we log it and then attempt to cleanup. There is a case here that we will likely be concerned with. if a `DuplicateName` occurs it is possible that a VM was here instead of a folder. in which case we should likely not call `remove` on the base folder. I do not currently see this as an issue as we have attempted a `Finder.Folder` read on the target and we did not get a folder back. which means we will try to make a folder. At that point we should fail and the cleanup should fail as well. Additionally if the folder did exist we would not be trying to make it. So I do not think the lack of check for a `DuplicateName` issue would be a problem here, though I could be missing some edge case that I have not though of. Feel free to add addtional thoughts and comments!